diff --git a/system/core/rootdir/init.rc b/system/core/rootdir/init.rc
--- a/system/core/rootdir/init.rc
+++ b/system/core/rootdir/init.rc
@@ -198,9 +198,6 @@ on post-fs-data
     chown system system /sys/class/android_usb/android0/f_rndis/ethaddr
     chmod 0660 /sys/class/android_usb/android0/f_rndis/ethaddr
 
-# Include extra init file
-    import /system/etc/init.local.rc
-
 on boot
 # basic network init
     ifup lo
