diff --git a/hardware/libhardware_legacy/wifi/wifi.c b/hardware/libhardware_legacy/wifi/wifi.c
index 119252b..2aa02db 100644
--- a/hardware/libhardware_legacy/wifi/wifi.c
+++ b/hardware/libhardware_legacy/wifi/wifi.c
@@ -160,6 +160,12 @@ static int insmod(const char *filename, const char *args)
     unsigned int size;
     int ret;
 
+    char command0[]="echo 0 > /sys/devices/platform/msm_sdcc.3/polling";
+    char command1[]="echo 1 > /sys/devices/platform/msm_sdcc.3/polling";
+    system(command1);
+    sched_yield();
+    usleep(1000000);
+
     module = load_file(filename, &size);
     if (!module)
         return -1;

